### DISCUSSION TITLE
supabase-js: How do I wait for supabase.auth.session() to return a value before continuing?

### DISCUSSION THREAD

**Discussion by silasabbott** [supabase/supabase#2315] [Questions]
I’m using the `handle()` function in SvelteKit’s `hooks.js` file that acts as middleware running before each page load/route change. I’m wanting to use this to check if the user is logged in and then redirect to login or dashboard accordingly. Since `supabase.auth.session()` isn’t a promise, however, my `handle()` function continues and incorrectly redirects the user even if they’re actually logged in. How do I wait for `supabase.auth.session()` to resolve before continuing to handle my routing?

My Javascript knowledge is spotty so forgive me if this is a simple answer!

**Comment by inian** [comment]
Hi @silasabbott, you are right in that `session` doesn't return a promise, but usually you listen to onAuthStateChange and redirect the user to the right component [like this](https://github.com/supabase/supabase/blob/a2dd30f6c8c14dff76b8755f8e5fe86f1a19257b/examples/react-user-management/src/App.js#L10-L16)

Does this example with SvelteKit help  - https://www.youtube.com/watch?v=j4AV2Liojk0

**Reply by silasabbott** [reply]
Thanks for the reply.

When does `onAuthStateChange` fire though? I thought it only fired on login or logout. I need it to fire and and return session or user data on initialization or page load.

**Reply by wilfredinni** [reply]
Hi, did you figure it out? Having the same issue with Vue

**Reply by lawrencecchen** [reply]
In svelte, you can abstract `session` into a svelte store. Here's what I typically use:
https://supabase-cheatsheet.vercel.app/docs/auth/svelte

Can't help you much with Vue though, sorry :)

**Reply by wilfredinni** [reply]
Thanks, I will have a look

**Reply by silasabbott** [reply]
@wilfredinni So, here's what I've gathered from asking around on the [Discord server](https://discord.supabase.com).

1. `supabase.auth.user()` and `supabase.auth.session()` initially return `undefined` when the user's logged-in state hasn't been gathered yet. Then, when the state has been retrieved, the value is set to either (1) `null` if the user is logged out or (2) the `user` or `session` object if the user is logged in with all the user/session info.
2. In an `async` function, to wait for for `supabase.auth.user()` or `supabase.auth.session()` to return its value before continuing, call the `supabase.auth.refreshSession()` method instead. This method returns a Promise, so you can use `await` or `.then().catch()` and grab the `{ error, user, session }` values from it. Here's an (untested) example of what that looks like:

``` js
async function handle () {
  const { error, user } = await supabase.auth.refreshSession()
  if (error) {
    // handle error
  } else if (user) {
    // redirect to dashboard
  } else {
    // redirect to login
  }
})
```

I have not tested this yet, so if someone has the chance before I do, please let us know if this works or not.

**Reply by JamesGarbaczewski** [reply]
I will definitely try that, here was my bad but working solution currently.
```js
let session = supabase.auth.session();
setTimeout(
	async () => {
		let session = supabase.auth.session();
                 // do stuff with session
	},
	session ? 1 : 1000
);
```

**Reply by MistaPidaus** [reply]
> I will definitely try that, here was my bad but working solution currently.
> 
> ```js
> let session = supabase.auth.session();
> setTimeout(
> 	async () => {
> 		let session = supabase.auth.session();
>                  // do stuff with session
> 	},
> 	session ? 1 : 1000
> );
> ```

thanks for sharing, @JamesGarbaczewski ! it works for me as temporary solution for now. I hope that Supabase team will add some kind of auth state/status so that we can render accordingly. i tried with promises and didn't work perfectly when token expired. 

thanks again!
